scalanet/src/io/iohk/scalanet/peergroup/kademlia/KPeerGroup.scala

```diff
@@ -0,0 +1,161 @@
+package io.iohk.scalanet.peergroup.kademlia
+
+import java.net.{InetAddress, InetSocketAddress}
+
+import io.iohk.scalanet.peergroup.PeerGroup.ServerEvent.ChannelCreated
+import io.iohk.scalanet.peergroup.PeerGroup.{ChannelSetupException, HandshakeException, ServerEvent}
+import io.iohk.scalanet.peergroup.kademlia.KPeerGroup.{ChannelImpl, NodeRecord, UnderlyingChannel}
+import io.iohk.scalanet.peergroup.{Channel, InetMultiAddress, PeerGroup}
+import monix.eval.Task
+import monix.execution.Scheduler
+import monix.reactive.Observable
+import monix.reactive.subjects.Subject
+import org.slf4j.{Logger, LoggerFactory}
+import scodec.bits.BitVector
+
+// PeerGroup using node records for addressing.
+// These node records are derived from Ethereum node records (https://eips.ethereum.org/EIPS/eip-778)
+class KPeerGroup[M](
+    val kRouter: KRouter[NodeRecord],
+    val server: Subject[ServerEvent[BitVector, M], ServerEvent[BitVector, M]],
+    underlyingPeerGroup: PeerGroup[InetMultiAddress, Either[NodeRecord, M]]
+)(implicit scheduler: Scheduler)
+    extends PeerGroup[BitVector, M] {
+
+  private val log = LoggerFactory.getLogger(getClass)
+
+  override def processAddress: BitVector = kRouter.config.nodeId
+
+  override def initialize(): Task[Unit] = {
+    // The protocol defined here requires that a peer
+    // connects then sends it node record as the first message
+    // (it could send earlier messages but they will be ignored)
+    // This node will then create a channel and reply with its
+    // own node record.
+    underlyingPeerGroup
+      .server()
+      .collect(ChannelCreated.collector)
+      .mergeMap(
+        channel =>
+          channel.in.collect {
+            case Left(nodeRecord) =>
+              acceptNodeRecord(channel, nodeRecord)
+          }
+      )
+      .subscribe()
+
+    Task.unit
+  }
+
+  override def client(to: BitVector): Task[Channel[BitVector, M]] = {
+    val underlyingChannelTask = Task
+      .fromFuture(kRouter.get(to)) // make the underlying kademlia lookup
+      .flatMap { record => // use the lookup's address info to obtain an underlying channel...
+        debug(s"Routing table lookup returns peer $record. Creating new channel.")
+        underlyingPeerGroup.client(InetMultiAddress(new InetSocketAddress(record.ip, record.tcp)))
```

Comment:
if we are using TCP, shouldn't we ask `underlyingPeerGroup` to have `TCPPeerGroup` type?
Also note that Kademlia actually uses UDP.